docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,526 @@
+
+With this approach of referenced resources, customers will be able to reference
+Api resource in Integration resource and apply both resource manifests
+together in their k8s cluster. This will allow customers to manage ACK custom
+resources using gitops methodologies. ACK controller will wait for API resource
+to get created and then use the ApiId to create Integration resource.
+
+The deletion strategy for the referenced resource is out of scope for this proposal
+right now. For Ex: If API resource is referenced in integration resource, the deletion
+of API resouorce will not be blocked. We plan to handle deletion strategy in next
+revision.
+
+## Scope
+
+### First Revision
+* Allow referencing resource from same api group in custom resource definitions
+* Allow referencing list of resources from same api group in custom resource
+definitions
+
+### Future Improvements
+* Add support for custom hooks while dereferencing resources
+* Add support for `selector` for referencing the objects
+* Add support for referencing resource from different api-groups
+* Add support for not deleting a resource if it is referenced in another resource
+* Add support for finite retries and terminal conditions when resolving references
+
+## Preferred Approach
+
+### High Level Overview
+
+* Introduce new construct in generator.yaml which will indicate to code-generator
+if a new spec field with type `AWSResourceReference` should be added to the custom
+resource definition.
+
+* This new spec field will work as additional input for an existing spec field.
+Addition of the extra spec field keeps the backward compatibility with existing
+custom resource definitions.
+
+* Validation will be present that only one of existing spec field OR new reference
+field is present in custom resource, not both.
+
+* If the existing spec field was marked as required, validation will be added that
+atleast one of existing spec field or new reference field is present.
+
+* If `AWSResourceReference` type field is present in desired state, first ACK
+runtime will query for the referenced resource and only when the specified field
+is found in referenced resource, the reconciliation loop will progress. Otherwise
+the request will get requeued for future processing.
+
+* Watchers for the referenced resource type will not be implemented in the first
+revision. The first version of this implementation will mostly be used for finding
+identifier of referenced resource and any update in referenced resource should not
+change it's identifier. As mentioned earlier, deletion will also be handled in
+coming revisions.
+
+* The code to determine when the referenced resource is synced will be generated,
+and additional hooks will be added for customization. Ex: Do not just check presence
+of a field but also check if the Status is available etc...
+
+### Low Level Implementation
+
+#### generator.yaml Changes
+
+```go
+type FieldConfig struct {
```

Comment:
I don't agree with this being a property of a field. I want to remove the need for service teams to choose names for the reference fields - in case every team chooses a different naming convention. I would greatly prefer that we append `Ref` to the end of every field we hope to override so that's it's absolutely clear to users that it's an overriding reference to another field.

eg. `apiID` becomes `apiIDRef`

Reply:
This would mean that we have a list of references defined per resource.

Reply:
I believe we can continue to have `ReferenceConfig` be part of `FieldConfig`, but can avoid your naming issues if we encourage the original field (e.g., `apiID`) to get the ReferenceConfig and not a new field. In other words, if this were the generator.yaml file for API Gateway v2:

```yaml
resources:
  Integration:
    fields:
      APIID:
        reference:
          from_resource: API
          from_path: Status.APIID
          to_field: APIID
```

then that would indicate that the `Integration.Spec.APIID` field could be populated with the `API.Status.APIID` field value of a resource referred to in a `Integration.Spec.APIIDRef` field, which we would automatically add to the `API` custom resource's schema when we found a `reference`. That way, we don't give anyone the ability to name their reference field something other than `<original field name>Ref`.

Reply:
> I don't agree with this being a property of a field.

+1, in addition to reasons above I think it also muddies the interface and use of `custom_field`. If you were to keep it in FieldConfig, then it'd be clearer to extend custom_field to have a `ref_of` type (list_of, map_of, ref_of) imo

Reply:
As per Jay's guidance, I am keeping the config inside fieldConfig but I have added another item for future to support references in custom fields